### DISCUSSION TITLE
How secure is Bun compared to Deno?

### DISCUSSION THREAD

**Discussion by renet** [oven-sh/bun#725] [General]
I've searched the GItHub issues and community, but didn't really find anything that told me about how Bun is thinking about and handling security. It's super fast - okay, but does that mean that security aspects are not that much in the focus of its developers (at least compared to Deno)? Is there any documentation about security concepts that I just didn't manage to find? Thanks a Bunch! 🥯

**Comment by Jarred-Sumner** [comment]
Rather than runtime permission checks (Deno's model) which could potentially have bugs that lead to permissions being ignored, Bun plans to have binary dead code elimination based on statically analyzing what features/native modules are used by the code.  

Instead of checking if a feature is allowed at runtime, it is safer to just not have potentially unsafe or undesired code in the binary to begin with. Can't have bugs in code that doesn't exist

A different build of Bun will exist for each feature along with tooling for choosing the exact build of Bun with allowed features included. This will probably be limited to an edge runtime build of Bun (for Linux) focused on production usage in servers.



This is not implemented (or started) yet, but that's the plan.

Bun might explore runtime permission checks in the future, but it just doesn't seem as sound as not including the code to begin with.

Additionally, Bun will undergo regular security audits once a little more stable

**Reply by trogers1** [reply]
Very cool idea! I know this is still in the 'concept' stage, but I'm curious: it seems like 'statically analyzing what feature/native modules are used by the code' would not identify if features and modules were being used that weren't the intention of the user, unless I'm completely misunderstanding the concept (totally possible 😅).

For example, if I download a random package for something simple, say csv-json conversion, but it starts making network calls in the background, that 'feature' (network access) _is_ being used by the code, even though I never intended it to be. How would this sort of paradigm catch this instance? With Deno, I believe that would be caught since those permissions weren't given at runtime (with `--allow-net`). Node, of course, wouldn't catch anything at all 😂

**Reply by kevlened** [reply]
> For example, if I download a random package for something simple, say csv-json conversion, but it starts making network calls in the background, that 'feature' (network access) _is_ being used by the code, even though I never intended it to be. How would this sort of paradigm catch this instance? With Deno, I believe that would be caught since those permissions weren't given at runtime (with `--allow-net`).

I _believe_ he addresses that here:

> A different build of Bun will exist for each feature along with tooling for choosing the exact build of Bun with allowed features included

In your example, to avoid network requests, you would use a version of bun that was compiled without network support.

**Reply by trogers1** [reply]
Ah! That makes sense. I like that idea a lot... Sorry for the dumb question, I just totally missed it 😂 It makes sense though. Very cool.

**Reply by samuba** [reply]
Very Interesting idea. But this approach means there is no granularity at all. 
With potentially infected packages in my app I want to make sure that it is not communicating to e.g. malicious servers or accessing /etc/passwd while still being able to communicate to my servers and accessing /temp/myApp.

**Reply by kevlened** [reply]
@samuba This is the same granularity deno offers. [Here's the original discussion on implementing module-level permissions in deno.](https://github.com/denoland/deno/issues/171) They decided against it, because the permissions are too difficult to enforce.

**Reply by samuba** [reply]
@kevlened Deno offers white/black listing for urls, file paths etc.
https://deno.land/manual@v1.36.3/basics/permissions

**Reply by kevlened** [reply]
@samuba Ah, I misunderstood. I assumed you meant your code could access /temp/myApp, while 3rd parties could not.

Deno's permissions _are_ more granular than the current scope of this proposal.

**Reply by matthew-dean** [reply]
This... answer doesn't really make sense. Deno, for example, allows you to evaluate any arbitrary untrusted JavaScript code from a third-party without, say, allowing file access. By restricting file access or file access only to supplied directories, you can provide things like a scripting language from user input that runs in server process, but which has no access to system resources. Dead code elimination and static analysis don't address this. Without a secure-by-default model (or at least an adjustable security model?), it makes Bun unusable for scenarios in which the JS / TS code cannot and should not be trusted.

**Reply by matthew-dean** [reply]
Unless Bun provides some kind of secure VM somehow that doesn't allow access to system resources or the process object etc etc?

**Reply by dzmitry-lahoda** [reply]
may be unix is way to go to run bun with restrictions at least on port/inode level via ebpf/jails?

would not ebpf modules and/or reverse proxy allow urls blocking?

**Reply by ronxjansen** [reply]
As someone who has first hand experienced a supply chain attack, I cannot stress enough the importance of granular permission levels - like Deno. 

Not including a `net` or `fs` module is not a solution. Basically every project I have worked with needs these modules and you could whitelist the URLs and directories they are allowed to access. 

Security is going to be more and more important in the years to come. For us it is already decisive to which stack we will use.

**Reply by lirantal** [reply]
+1

I've researched Bun's security posture and have put that together at https://www.bunsecurity.dev/
Did you by any chance see that before?

**Reply by ronxjansen** [reply]
@lirantal I did, although I feel "Bun secure-by-default" claim is at least exaggerated. One could also argue that this claim is misleading or false. Bun does not secure anything by default, other then not running pre/post install scripts. To my knowledge, the entire runtime is just as exposed as the Node.js runtime.

**Reply by lirantal** [reply]
I see what you mean and can relate. There are some aspects of Bun which demonstrate the "secure by default" paradigm (as you called out the pre/post install scripts is one of them), another is the "secure by design" paradigm around APIs.

Are there specific topics you'd expect the focus to be driven towards?

**Reply by JaykeOps** [reply]
I couldn't hold myself from pointing this out, but in today's threat landscape security relies on defense in depth. The argument against a runtime permission model, because it might contain bugs is like arguing that a door lock might ship from the factory with a mechanical defect, therefore no door locks at all.

Whether it is containers, VM's etc, no single security boundary is 100% flawless. We use them anyway because security is about layering friction, not about perfection.

Dead Code Elimination is a whole different story. It is a good, it protects you from the code you don't need. But it does not protect you from the code you do need, like runtime containment does.

**Comment by lirantal** [comment]
Not entirely on this topic of permissions but I am building new Bun Security course and am interested to learn and hear what other topics and domains would be of interest to teach and educate on (specifically to Bun but also anything adjacent that makes sense).

**Reply by matthew-dean** [reply]
I think the first thing I would cover is that Deno has security by default and Bun, by default, is less secure.

**Reply by lirantal** [reply]
I haven't researched Deno so I can't genuinely comment about its security posture but Bun is taking some very proactive and smart secure by default and secure by design approach and very much to my liking.

**Reply by matthew-dean** [reply]
Oh? Did they add security since this thread? Like, do scripts now not have file / network etc access by default?

**Reply by lirantal** [reply]
There are other forms of security?

**Comment by Jobians** [comment]
Any updates?

**Comment by zuohuadong** [comment]
Any updates?

**Reply by Jobians** [reply]
Deno is significantly more secure by default than Bun. Deno was built from the ground up with a "secure-by-default" philosophy, whereas Bun prioritizes raw performance and compatibility with the Node.js ecosystem, adopting a more permissive model.

**Reply by dzmitry-lahoda** [reply]
> Any updates?

It may appear that one should start bun within container (not OCI, but container of some sort) which is used by modern AI tools, and elevate access as they do.
